### PR TITLE
CP-346: Create 'recommended' preset for ESlint (w/ ~ current FE monorepo standards) in the open-turo repo

### PR DESCRIPTION
**Pull request by romenrg**
**Description**

The detailed description referencing some Turo repositories is in our internal Jira system ([see particular comment](https://team-turo.atlassian.net/browse/CP-346?focusedCommentId=183612)).

This PR includes some changes to the default preset, including changes in the `test` directory structure to follow our internal conventions. This causes a breaking change.

Also, this PR creates a new `legacy.js` file, which includes configuration for our front-end projects.
Even though the default preset remains to be the existing one as the `index.js`, the new `legacy` preset can also be extended from projects that depend on this package, as noted in the README.

**Breaking change**: as noted above, this PR includes a breaking change. Repositories currently extending this `eslint` config package are expected to follow the `/test` directory structure convention, rather than the previous `src/**/__tests__/**`

Fixes [#CP-346](https://team-turo.atlassian.net/browse/CP-346)

**Changes**

* refactor: current eslint conf to 'strict'
* feat: default to new recommended settings
* docs: added guidelines to use one of the two configs

🚀 PR created with [fotingo](https://github.com/tagoro9/fotingo)

### CODE REVIEW

**Review by Toolo — APPROVED** [review]
LGTM 😄

File: index.js [hunk 1]
```diff
@@ -1,10 +1,13 @@
 module.exports = {
   root: true,
   env: {
+    browser: true,
```
Comment (tagoro9) [inline index.js:4]: I don't think this should reference browser. I see there are a couple of rules here targeting react / front-end. For those cases we should add the rules to eslint-config-react and extend from there as well (or have a dedicated preset for react that extends the default one and archive the other repo)

File: strict.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        allowModules: ["@jest/globals"],
+      },
+    ],
+    "node/no-unsupported-features/es-syntax": "off",
+    "node/no-missing-import": "off",
+    "simple-import-sort/imports": "error",
+    "simple-import-sort/exports": "error",
+    "sort-destructure-keys/sort-destructure-keys": "error",
+    "@typescript-eslint/no-unused-vars": [
+      "error",
```
Comment (tagoro9) [inline strict.js:61]: It feels like some rules in the strict mode should be part of the default preset like this one

File: README.md [hunk 3]
```diff
@@ -10,12 +10,34 @@ Install the package and all of its peer dependencies:
 npx install-peerdeps --dev @open-turo/eslint-config-typescript
 ```
 
-Then in your `.eslintrc` file add:
+Then in your `.eslintrc` file extend from one of the two configurations included in this package:
+
+1. The default, recommended for most Turo front-end projects
+2. The strict, that was being used for some node projects
```
[the remark this answers is not in the document]
Comment (tagoro9) [reply README.md:16]: Should we still reverse the logic and keep the strict the default (with some required updates for that to be possible) and have a relaxed preset for legacy projects?

**Review by tagoro9 — COMMENTED** [review]
You probably need to redo your commits. I don't think a refactor commit works with breaking changes. Also a couple of thoughts and questions around the configuration, which should be the default one and to decouple things from the react (if I want to create something like a lambda I should be able to use the default preset from here)

File: strict.js (hunk 2 above)
Comment (romenrg) [reply strict.js:61]: It already is: https://github.com/open-turo/eslint-config-typescript/pull/118/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R81

Comment (tagoro9) [reply strict.js:61]: I see, I missed that. Then what we could do is have one preset extend from the other to avoid a lot of repetition and have a better understanding of the things one preset is adding

**Comment by romenrg** [comment]
@tagoro9 Based on some of your comments, it seems you prefer not to have web-related rules / config in this repo. And I can see your point for scenarios like node applications that don't have anything to do with the web.

I definitely agree with the react-specific ones. However, there is a gray area with some "browser"-related ones. Someone could argue it is harmless to include some of them and it is a common enough use case to consider them to be included. I can also imagine that a similar claim could be made from the opposite standpoint: there are node-specific rules that don't make sense in a web environment, while both being valid Typescript projects.

Thinking a bit more deep into the idea of moving web-related config to the `eslint-config-react` reo, it doesn't feel right either, since you can have web applications that don't use React, but other frameworks (or no frameworks). Following that train of thought, that leads me to think that if we were to go that route, we would either have to create an intermediate framework-agnostic repo (`eslint-config-typescript-web`) that is then extended by `eslint-config-react` for React. Or alternatively, we would rename the `eslint-config-react` to `eslint-config-typescript-web` and couple the two concepts.

I lean more towards option 1: creating a separate `eslint-config-typescript-web` and then extending it in a ``eslint-config-typescript-react`. Although I can see arguments against having so many levels.

**Comment by tagoro9** [comment]
@romenrg 

> eslint-config-react reo, it doesn't feel right either, since you can have web applications that don't use React, but other frameworks (or no frameworks). Following that train of thought, that leads me to think that if we were to go that route, we would either have to create an intermediate framework-agnostic repo (eslint-config-typescript-web

I agree with you on this, but being practical and looking at what we use today, we only use react as a framework, so adding that extra layer may not be needed yet and would add a bit of extra maintenance burden.

> eslint-config-typescript-react

Is there any particular react config that is specific to typescript or would all the things could be added to just `eslint-config-react`?

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (romenrg) [reply README.md:16]: > This is confusing. Why shouldn't be using the strict mode in any non legacy project?

This issue goes back to the tension I see between node and web projects.
The `strict` mode includes stuff that is node specific and/or is too rigid for our existing codebase (e.g. `unicorn`), meaning that we won't be able to use it for our current efforts on monorepo standards.
That's why we discussed creating a "recommended" eslint config that works for our front-end repositories.
It felt that our front-end projects are the main consumers of this `eslint` config, and in that context it made more sense to have that as the default

File: index.js [hunk 4]
```diff
@@ -1,10 +1,13 @@
...
+    es6: true,
+    jest: true,
     node: true,
-    es2022: true,
   },
   parser: "@typescript-eslint/parser",
+  ignorePatterns: ["babel.config.js", "jest.config.js"],
```
[the remark this answers is not in the document]
Comment (romenrg) [reply index.js:10]: In this particular case we are just telling `eslint` to ignore babel files when applying the linting to the project. Don't see harm in having those here (as in a gitignore we typically have more files / file patterns that we have in the project), but no strong opinion and can move to the web-specific repo that extends this one.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (tagoro9) [reply README.md:16]: I see. Outside of existing projects, where bringing up to these standards would be complex, what issues do you see in having strong rules like unicorn in new projects? Or is it really too strict?

[the remark this answers is not in the document]
Comment (tagoro9) [reply README.md:16]: In my mind I'm thinking of existing projects as legacy and we should try to move them to the new world eventually and thus the new preset, to support them until we can afford to do that.

File: strict.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
+module.exports = {
+  root: true,
```
[the remark this answers is not in the document]
Comment (romenrg) [reply strict.js:2]: And actually, I'm thinking we could follow the same approach for `web` and `react` presets, making them just be files in this repo; so we avoid having multiple repos and then having a `web` and `react` hierarchy is less of a concern.

Will have to explore how to extend / apply the hierarchy within the same repo between JS files

File: strict.js (hunk 2 above)
Comment (romenrg) [reply strict.js:61]: Agreed. Although will have to explore how to create that hierarchy between JS files, as I'm assuming the usual `extends` won't be applicable here as a file reference (but maybe it works).

Related: https://github.com/open-turo/eslint-config-typescript/pull/118/files#r1025669400

File: index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (tagoro9) [reply index.js:10]: Maybe worth thinking, why are we telling eslint to ignore these files? Couldn't we lint them like any other? 

If that is not possible, I agree with you that having them here would not hurt, but maybe instead we could have a `.eslintignore` file in each project?

File: strict.js (hunk 2 above)
Comment (tagoro9) [reply strict.js:61]: I think it should probably work fine but not completely sure, so we should give it a try

File: strict.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (tagoro9) [reply strict.js:2]: > making them just be files in this repo

I think this would generate a problem with dependencies and peer dependencies. Probably the `react` preset will have peer dependencies that are not needed by the node preset.

That is why we decided to split things into separate repos, so we could better control peer dependencies. 

With our internal modules we ended up having to install `react` in a node.js project that didn't needed and that was very confusing.

**Comment by romenrg** [comment]
> eslint-config-typescript-react

> Is there any particular react config that is specific to typescript or would all the things could be added to just eslint-config-react?

I thought that, since we the react config would be extending the `eslint-config-typescript` preset, it should be reflected in the repo name. However, I'm also thinking we could keep the name as `eslint-config-react` and have 2 files within it, one for JS (eslint-config-react/js) and one for TS (eslint-config-react/ts), extending the corresponding top level eslint config. And one could be the default (not sure if nor which one)

**Comment by tagoro9** [comment]
> However, I'm also thinking we could keep the name as eslint-config-react and have 2 files within it, one for JS (eslint-config-react/js) and one for TS (eslint-config-react/ts), extending the corresponding top level eslint config in each case. Plus, one could be the default (not sure, nor which one in case we want one default)

I think we can start there. We may hit the same issue with peer dependencies if there is any peer dep that the TS preset needs that is not needed for the JS preset, but we can see then

**Comment by tagoro9** [comment]
@romenrg ☝️

File: index.js [hunk 6]
```diff
@@ -64,6 +85,20 @@ module.exports = {
...
     ],
+    "@typescript-eslint/restrict-template-expressions": [
+      "error",
+      {
+        allowBoolean: true,
+      },
+    ],
+    "@typescript-eslint/unbound-method": "error",
+    "@typescript-eslint/no-unnecessary-type-assertion": "error",
+    // All the following TS-related rules turned to warnings are caused by the module not being "TS native"
```
[the remark this answers is not in the document]
Comment (romenrg) [reply index.js]: I've been thinking about this, and this seems to be another example of a tricky balance between this repo and the 'recommended' configuration we have come up with after FE-extractions work internally.

I've redone the message in a new commit with some context: https://github.com/open-turo/eslint-config-typescript/pull/118/commits/9e952a1fde9ed7ffd35ad3904d015c69a074955c.

I find it necessary to keep this note and to explain it; and I would like to keep it here, as creating an internal repository for FE packages only for this reason feels unnecessary painful?

File: strict.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    ecmaVersion: "latest",
+    project: "./tsconfig.json",
+    sourceType: "module",
+  },
+  rules: {
+    "import/default": "off",
+    "import/named": "off",
+    "import/namespace": "off",
+    "import/no-default-export": "error",
+    "import/no-extraneous-dependencies": [
```
[the remark this answers is not in the document]
Comment (romenrg) [reply strict.js:43]: Addressed: https://github.com/open-turo/eslint-config-typescript/pull/118/commits/b82cc69c7263f0b749cf8d94937c24d6be52f38a

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (romenrg) [reply README.md:16]: > to support them until we can afford to do that

From my perspective, all the FE repositories, including the new ones created as part of extractions from the monorepo, will probably never use the `strict` preset, as it is currently configured.

If we don't feel comfortable having the front-end-compatible preset as the default, I can revert it; but it will force us to pick the right name for that preset. Maybe `stable`? And it will also mean we will have to change all the front-end packages `.eslintrc` to point to that other file.

I can do that. CC'ing @Toolo for awareness on this

**Comment by Toolo** [comment]
We could also impose the following restriction: React requires TS. This should remove the complexity of supporting both JS and TS files within the React config folder.

For existing projects (web-iframes-app only, I believe), I think we should just convert them to TS as that's the path we've set moving forward 💡

**Comment by tagoro9** [comment]
> We could also impose the following restriction: React requires TS [...]

@Toolo I think this is a fair assumption to make. I don't see a world where we would write JS + react

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (tagoro9) [reply README.md:16]: > will probably never use the strict preset, as it is currently configured.

What are the things / rules that are blocking this? If it's just unicorn, then let's see how we can relax it a bit

I think ideally achieving a place where both worlds meet would be great?

**Review by tagoro9 — COMMENTED** [review]
I think you probably haven't addressed all the previous comments.

If you could list what the problematic rules where with the original preset in the front-end projects, I think we can try to relax unicorn rules (assuming it's unicorn to a place where it works for both worlds)

Also, looking at your commit messages, I think you need to update the breaking change notes, as they are pointing to a private repo

```
BREAKING CHANGE: repos using this config have to modify the extends field in their .eslintrc file.
 We'll have to add the '/strict' suffix to the path.
 See example in spike / test PR: https://github.com/turo/typescript-template/pull/237/files
```

File: index.js (hunk 1 above)
Comment (romenrg) [reply index.js]: Removed here: https://github.com/open-turo/eslint-config-typescript/pull/118/commits/5aa24f91e46f3f2717aea032e4c97357e756228f
And added here: https://github.com/open-turo/eslint-config-react/pull/85

File: index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (romenrg) [reply index.js]: Also removed here the babel-specific one: https://github.com/open-turo/eslint-config-typescript/pull/118/commits/5aa24f91e46f3f2717aea032e4c97357e756228f
And added it in the react-specific repo here: https://github.com/open-turo/eslint-config-react/pull/85

File: index.js [hunk 8]
```diff
@@ -28,28 +31,46 @@ module.exports = {
...
     "import/prefer-default-export": "off",
     "jest/no-jest-import": "off",
     "node/no-unpublished-import": [
       "error",
       {
-        allowModules: ["@jest/globals"],
+        allowModules: [
+          "@jest/globals",
+          "@testing-library/dom",
+          "@testing-library/react",
```
[the remark this answers is not in the document]
Comment (romenrg) [reply index.js]: Also removed here: https://github.com/open-turo/eslint-config-typescript/pull/118/commits/5aa24f91e46f3f2717aea032e4c97357e756228f
And added here: https://github.com/open-turo/eslint-config-react/pull/85

File: index.js [hunk 9]
```diff
@@ -28,28 +31,46 @@ module.exports = {
...
+          "@testing-library/user-event",
+          "nock",
```
[the remark this answers is not in the document]
Comment (romenrg) [reply index.js]: Removed here: https://github.com/open-turo/eslint-config-typescript/pull/118/commits/5aa24f91e46f3f2717aea032e4c97357e756228f
And added here: https://github.com/open-turo/eslint-config-react/pull/85. Since this `react` repo is very tied to our front-end repositories config, it felt right to keep it there

File: strict.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (romenrg) [reply strict.js]: Update after all the final changes:

- The 2 presets that remain in the repo (defaul vs legacy) have `root: true` line, as they don't extend each other.
- On the other hand, we have moved all the React-related config to the `eslint-config-react` repository: https://github.com/open-turo/eslint-config-react/pull/85

File: strict.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (romenrg) [reply strict.js]: This commet referred to: https://github.com/open-turo/eslint-config-typescript/pull/118/commits/74cc6bbdf33fa4bb204859287edfffb01a511e5a

File: strict.js (hunk 2 above)
Comment (romenrg) [reply strict.js]: The `extends` line does work internally, just with the path to the file.

It is not applicable to this PR, though, as the two presets are considered separate.

It does apply to the `eslint-config-react` PR, where an example can be found: https://github.com/open-turo/eslint-config-react/pull/85

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (romenrg) [reply README.md]: Update:

We ended up keeping the original preset as the default and added a `legacy` one based on our front-end config.

The README has been updated accordingly in: https://github.com/open-turo/eslint-config-typescript/pull/118/commits/446ea9d4fb034282fb489bca9992c8c5533a439e

**Comment by romenrg** [comment]
**Final update**

After a conversation with @tagoro9 yesterday, I've re-done commits here to reflect the final agreements.

I've updated both the PR description and the commit messages so that it is clear to review this PR now. I have also added follow-up comments to the conversations we had here, but I would recommend new reviews to skip them, as they are all now addressed and commit history has changed since those conversations were relevant.

**So, in summary: for those reviewing now, I would recommend to simply read the updated PR description and then review changes commit by commit.**

**Review by Toolo — APPROVED** [review]
This looks reasonable to me: the default remains unchanged and we have a slightly better `legacy.js` to use in the rest of projects. 👍

**Review by NGorco — APPROVED** [review]
